### PR TITLE
Add unit tests to azuremanagedcontrolplane reconciler

### LINKED ISSUE #4290: Create unit tests for `azuremanagedcontrolplane_reconciler.go`
/kind cleanup

- [ ] Add unit tests to `azuremanagedcontrolplane_reconciler.go` 

This issue is part of a larger issue #3649 to help with coverage and ensure confidence

### PR DESCRIPTION
**Pull request by troy0820**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This will increase coverage around the controller/reconciler code.  
 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4290 

**Special notes for your reviewer**:
This test is marked as an error happening for the following reasons.  When it hits the path to `StoreClusterInfo` it cannot create a remote client because that call is trying to reach out to do this.  This is expected and not meant to derail the effort around coverage.  However, this doesn't show the successful path around the `reconcileNormal`.  There are many functions/methods that I had to extract and refactor to become testable to reduce the level of abstraction around how nested the `New` funcs are within certain places where I can't inject but patch as done in a previous PR and as well as done here.   

I did create new mocks to test the clients that we export in the services package `AzureClient` which held azure-sdk-for-go clients and added coverage around this as well.  
I left the commits as a single point of reference if you want to review this within the context of the commits, however I can squash if this meets the satisfaction of the maintainers.  

Also, the test for `CreateOrUpdateAsync` end in error (not failing) because the poller that is returned is something that I cannot inject to `EXPECT` on the return value.  The call cannot be done because the return value for runtime.Poller[<pick your response>] is not something that the azure-sdk-defined and cannot make that interface return value a thing to utilize in the test.  

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
```release-note
None
```

### CODE REVIEW

File: controllers/azuremanagedcontrolplane_reconciler.go [hunk 1]
```diff
@@ -150,7 +150,6 @@ func (r *azureManagedControlPlaneService) reconcileKubeconfig(ctx context.Contex
 			return errors.Wrap(err, "failed to reconcile kubeconfig secret for cluster")
 		}
 	}
-
```
Comment (troy0820) [inline controllers/azuremanagedcontrolplane_reconciler.go:153]: I can add this line back in.

File: azure/services/subnets/client.go [hunk 2]
```diff
@@ -28,13 +28,23 @@ import (
...
 type AzureClient struct {
-	subnets *armnetwork.SubnetsClient
+	subnets SubnetsClient
 }
 
+var _ SubnetsClient = (*armnetwork.SubnetsClient)(nil)
+
 // NewClient creates a new subnets client from an authorizer.
-func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+var NewClient = func(auth azure.Authorizer) (*AzureClient, error) {
```
Comment (troy0820) [inline azure/services/subnets/client.go:47]: I can remove this because we are not using this to patch anything.

File: azure/services/subnets/client.go [hunk 3]
```diff
@@ -43,7 +53,7 @@ func NewClient(auth azure.Authorizer) (*AzureClient, error) {
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
 	}
-	return &AzureClient{factory.NewSubnetsClient()}, nil
+	return &AzureClient{subnets: factory.NewSubnetsClient()}, nil
```
Comment (troy0820) [inline azure/services/subnets/client.go:56]: I can also remove this.

File: controllers/azuremanagedcontrolplane_reconciler.go (hunk 1 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_reconciler.go]: Added back, resolving.

File: azure/services/subnets/client.go (hunk 2 above)
Comment (troy0820) [reply azure/services/subnets/client.go]: Removed, resolving

File: azure/services/subnets/client.go (hunk 3 above)
Comment (troy0820) [reply azure/services/subnets/client.go]: Removed, resolving.

**Review by willie-yao — COMMENTED** [review]
Thanks for your great work on this @troy0820! This is some really good stuff, but I might have a comment on the approach to this PR. This PR modifies many of the Azure services, which are in the middle of being refactored to use ASO in #3402 rather than the Azure SDK for Go. I'm not entirely sure if the same logic would translate to the new way of structuring services, as services will no longer have their own CreateOrUpdate functions.

Also, I'm not sure if those changes are entirely necessary. If the goal is to pass in scopes to `reconcileNormal`, you can create them manually like how the azuremachine reconciler test does it: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremachine_controller_test.go#L338

This is definitely my bad as I misunderstood your proposal to use mocks, so I'm super sorry for misleading you! Do you think the approach used in the `azuremachine_controller_test.go` would work for your case?

File: azure/services/tags/mock_tags/doc.go [hunk 4]
```diff
@@ -18,6 +18,7 @@ limitations under the License.
 //
 //go:generate ../../../../hack/tools/bin/mockgen -destination client_mock.go -package mock_tags -source ../client.go Client
 //go:generate ../../../../hack/tools/bin/mockgen -destination tags_mock.go -package mock_tags -source ../tags.go TagScope
+//go:generate ../../../../hack/tools/bin/mockgen -destination ta -package mock_tags -source ../tags.go TagsClient
```
Comment (CecileRobertMichon) [inline azure/services/tags/mock_tags/doc.go:21]: > ta

is this a typo?

Comment (troy0820) [reply azure/services/tags/mock_tags/doc.go:21]: This is indeed a typo.  I'm not sure how it got generated

**Comment by troy0820** [comment]
> Thanks for your great work on this @troy0820! This is some really good stuff, but I might have a comment on the approach to this PR. This PR modifies many of the Azure services, which are in the middle of being refactored to use ASO in #3402 rather than the Azure SDK for Go. I'm not entirely sure if the same logic would translate to the new way of structuring services, as services will no longer have their own CreateOrUpdate functions.
> 
> Also, I'm not sure if those changes are entirely necessary. If the goal is to pass in scopes to `reconcileNormal`, you can create them manually like how the azuremachine reconciler test does it: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremachine_controller_test.go#L338
> 
> This is definitely my bad as I misunderstood your proposal to use mocks, so I'm super sorry for misleading you! Do you think the approach used in the `azuremachine_controller_test.go` would work for your case?

This doesn't change any of the logic that exists but makes the `azure-sdk-for-go` clients an interface to allow this to be tested and not reach out to azure to do things.  Because this is migrating to ASO, it may not make sense to change this while allowing that other work to continue.  This doesn't offer any benefit besides covering some lines and allowing to utilize these methods (the top level of abstraction) within the codebase to test.  **I can remove them, including the mocks.**

As far as the test in question, I would really only need to do the patch method for the `new....` function that I already have there.

I didn't want to use the mock for the type so I can follow the path in the method `r.reconcileKubeconfig`. It only gets stuck because the scopes has to call the `r.StoreClusterInfo()` and not being able to get around the `newRemote` method inside that function without patching makes the test fail/error (which is supposed to).  Which is fine unless I used the mock and using the mock removes the setup for the test I'm doing.  

TLDR; I will remove the client mocks and test for the `azure/services/<service_name>/client_test.go` and go as far down the code path as I can with the mocks I have in place.

File: azure/services/tags/mock_tags/doc.go (hunk 4 above)
Comment (troy0820) [reply azure/services/tags/mock_tags/doc.go]: Removing and resolving.

File: azure/services/virtualmachines/client.go [hunk 5]
```diff
@@ -29,19 +29,17 @@ import (
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
-type (
-	// AzureClient contains the Azure go-sdk Client.
-	AzureClient struct {
-		virtualmachines *armcompute.VirtualMachinesClient
-	}
+// AzureClient contains the Azure go-sdk Client.
```
Comment (willie-yao) [inline azure/services/virtualmachines/client.go:32]: Was this diff intended? It looks like it's just changing how the type is wrapping the structs

File: controllers/azuremanagedcontrolplane_controller.go [hunk 6]
```diff
@@ -237,8 +237,7 @@ func (amcpr *AzureManagedControlPlaneReconciler) reconcileNormal(ctx context.Con
 			return reconcile.Result{}, err
 		}
 	}
-
-	svc, err := newAzureManagedControlPlaneReconciler(scope)
+	svc, err := getNewAzureManagedControlPlaneReconciler(scope)
```
Comment (willie-yao) [inline controllers/azuremanagedcontrolplane_controller.go:240]: nit: preserve the newline unless it's better style-wise

**Review by willie-yao — COMMENTED** [review]
Looks good pending a few nits! Thanks for your patience here and I apologize again that the mocks didn't work out. I think we can still use them as a reference though in the future once ASO is implemented!

File: azure/services/virtualmachines/client.go (hunk 5 above)
Comment (troy0820) [reply azure/services/virtualmachines/client.go:32]: I think this may be left over from when I had the other interfaces,  I can change it back if it's more readable.

File: controllers/azuremanagedcontrolplane_controller.go (hunk 6 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller.go:240]: This may have been an editor thing.  I'll remove the newline.

Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller.go:241]: Removed and resolving.

File: controllers/azuremanagedcontrolplane_controller_test.go [hunk 7]
```diff
@@ -176,3 +183,80 @@ func TestAzureManagedControlPlaneReconcilePaused(t *testing.T) {
...
+
+		peService := &privateendpoints.Service{
+			Scope:      privateEndpointsClient,
+			Reconciler: mock_async.NewMockReconciler(ctrlr),
+		}
+		return &azureManagedControlPlaneService{
+			kubeclient: scopes.Client,
+			scope:      scopes,
+			services: []azure.ServiceReconciler{
+				peService,
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_controller_test.go:238]: Could we use a `*mock_azure.MockServiceReconciler` here instead of a concrete implementation of a particular service?

File: controllers/azuremanagedcontrolplane_controller.go [hunk 8]
```diff
@@ -342,3 +342,5 @@ func (amcpr *AzureManagedControlPlaneReconciler) ClusterToAzureManagedControlPla
 
 	return nil
 }
+
+var getNewAzureManagedControlPlaneReconciler = newAzureManagedControlPlaneReconciler
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_controller.go:346]: In case we ever run tests in parallel that each modify this function, I could see having a single variable like this causing race conditions. Is there a way we can scope these implementation details to an individual instance of something like `AzureManagedControlPlaneReconciler` like we do for the AzureMachine tests?

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/454d632e21446c5796797bb8ff03977929ecb7e0/controllers/azuremachine_reconciler.go#L107-L109

File: controllers/azuremanagedcontrolplane_controller_test.go [hunk 9]
```diff
@@ -176,3 +183,80 @@ func TestAzureManagedControlPlaneReconcilePaused(t *testing.T) {
...
+		Spec:   infrav1.AzureManagedControlPlaneSpec{},
+		Status: infrav1.AzureManagedControlPlaneStatus{},
+	}
+	scheme, err := newScheme()
+	g.Expect(err).ToNot(HaveOccurred())
+
+	client := fake.NewClientBuilder().WithScheme(scheme).WithObjects(cp).WithStatusSubresource(cp).Build()
+	amcpr := &AzureManagedControlPlaneReconciler{
+		Client:   client,
+		Recorder: record.NewFakeRecorder(42),
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_controller_test.go:204]: Is there a reason behind this particular channel buffer size of 42? Or is this even necessary at all if the test doesn't read recorded events?

Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:204]: This doesn't even seem necessary.  The arbitrary 42 was chosen for no particular reason.  I can remove this line as it's not necessary.

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 7 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:238]: @nojnhuh I believe if we do that, it will not walk the code path being tested because the mock will `EXPECT()` and `Return` whatever we like to obfuscate the functionality.  

We can use that and still provide coverage but it will lack the depth because that abstraction doesn't push through the intention of that test.  

If it's easier to do that, I can change this to match the suggestion with the understanding that this abstraction will not provide the confidence around changes to the functionality around services needing to "Reconcile" in it's own right. 

However we can move forward, I'm okay with.

File: controllers/azuremanagedcontrolplane_controller.go (hunk 8 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller.go:346]: I was thinking about this and maybe deferring it back to the actual implementation however, that doesn't allow us being able to catch that when parralell uses this package level variable to "patch".  

If we make this part of the actual reciever 
```go
func (amcpr *AzureManagedControlPlaneReconciler) getNewAzureManagedControlPlaneReconciler = newAzureManagedControlPlaneReconciler
```
we may be in better shape which mimics what you have above.  I can make these changes.

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 9 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go]: I removed this as this doesn't do anything for the test. Resolving.

File: controllers/azuremanagedcontrolplane_controller.go (hunk 8 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller.go]: Instead of making it part of the reciever, I took the liberty to do the following: 
```go
type AzureManagedControlPlaneReconciler struct {
     client.Client
     Recorder                                 record.EventRecorder
     ReconcileTimeout                         time.Duration
     WatchFilterValue                         string
     getNewAzureManagedControlPlaneReconciler func(scope *scope.ManagedControlPlaneScope) (*azureManagedControlPlaneService, error)
   }
```
On instantiation it gets set to: 

```go
amcpr.getNewAzureManagedControlPlaneReconciler = newAzureManagedControlPlaneReconciler
```

Testing allows us to rescope this as demonstrated in this PR.

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 7 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:238]: Let me know which way you'd rather me go with and I can make the changes @nojnhuh

Comment (nojnhuh) [reply controllers/azuremanagedcontrolplane_controller_test.go:238]: Which particular code path do you mean would no longer be exercised? Using a mock would skip over calling the real privateendpoints `Reconcile()`, but since that's already tested in isolation, I think it's acceptable to hide those details at this level of abstraction for the sake of testing `reconcileNormal()`.

File: controllers/azuremanagedcontrolplane_controller.go (hunk 8 above)
Comment (nojnhuh) [reply controllers/azuremanagedcontrolplane_controller.go]: SGTM, thanks!

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 7 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:238]: Perfect, I'll make the necessary changes and apply the squashing.

File: controllers/azuremanagedcontrolplane_controller_test.go [hunk 10]
```diff
@@ -176,3 +182,82 @@ func TestAzureManagedControlPlaneReconcilePaused(t *testing.T) {
...
+
+	amcpr.getNewAzureManagedControlPlaneReconciler = func(scope *scope.ManagedControlPlaneScope) (*azureManagedControlPlaneService, error) {
+		ctrlr := gomock.NewController(t)
+		defer ctrlr.Finish()
+		svcr := mock_azure.NewMockServiceReconciler(ctrlr)
+		svcr.EXPECT().Reconcile(gomock.Any()).Return(nil).AnyTimes()
+
+		return &azureManagedControlPlaneService{
+			kubeclient: scopes.Client,
+			scope:      scopes,
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_controller_test.go:237]: I know these have the same value, but would it technically be more correct to refer to the `scope` parameter instead of `scopes` here? Otherwise I see `scope` isn't used.

File: controllers/azuremanagedcontrolplane_controller_test.go [hunk 11]
```diff
@@ -176,3 +182,82 @@ func TestAzureManagedControlPlaneReconcilePaused(t *testing.T) {
...
+		ControlPlane: cp,
+	}
+	scopes.SetAdminKubeconfigData(createFakeKubeConfig())
+	scopes.SetUserKubeconfigData(createFakeKubeConfig())
+
+	amcpr.getNewAzureManagedControlPlaneReconciler = func(scope *scope.ManagedControlPlaneScope) (*azureManagedControlPlaneService, error) {
+		ctrlr := gomock.NewController(t)
+		defer ctrlr.Finish()
+		svcr := mock_azure.NewMockServiceReconciler(ctrlr)
+		svcr.EXPECT().Reconcile(gomock.Any()).Return(nil).AnyTimes()
```
Comment (nojnhuh) [inline controllers/azuremanagedcontrolplane_controller_test.go:233]: I noticed that this test fails if I remove the `.AnyTimes()` here due to not calling `Reconcile()` the expected one time, which is wrong because I think the `ctrlr.Finish()` is executing too soon (at the end of this inner function vs. at the very end of the test).

I'd vote to remove the `AnyTimes()` or make an equivalent change that ensures we're calling that `Reconcile` at least once, and then fix the test failure by removing `defer ctrlr.Finish()` entirely, which isn't necessary when initializing the mock controller with a `*testing.T` like is already being done here.

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 7 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go]: This still fails because of the `reconcileKubeConfig` method at the end of `StoreClusterInfo` but removes the addition of the added serviceReconciler, etc.  

This should be good to go. Let me know what you think @nojnhuh

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 11 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:233]: _You can tell I used to use the old gomock when writing tests, that I forgot_ (facepalm) 

Removing and pushing again.

File: controllers/azuremanagedcontrolplane_controller_test.go (hunk 10 above)
Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:236]: I'm not following exactly what you are referring to.  The scopes needs to carry the client, controlplane, etc.  Are you referring to the kubeclient? which is a client.Client?

Comment (nojnhuh) [reply controllers/azuremanagedcontrolplane_controller_test.go:236]: I mean the `scope` parameter to this function instead of the `scopes` variable defined outside this function.

Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go:236]: Yeah that makes sense, and would much rather have it reflect the value passed in (which is scopes) but get it's value from the param in the func we are patching.  

Nice catch.  Thanks for that @nojnhuh

Comment (troy0820) [reply controllers/azuremanagedcontrolplane_controller_test.go]: Updated and pushed.  

Now we should be good to go @nojnhuh

**Review by nojnhuh — COMMENTED** [review]
/lgtm

Thanks @troy0820!
